src/main/java/com/google/api/codegen/go/GoGapicContext.java

```diff
@@ -167,12 +169,24 @@ public String getServiceClientConstructorName(Interface service) {
   /**
    * Returns the local name used in Go files for the package of the proto.
    */
-  private static String localPackageName(ProtoElement proto) {
+  private String localPackageName(ProtoElement proto) {
     String goPackage = proto.getFile().getProto().getOptions().getGoPackage();
-    if (goPackage.startsWith(CORE_PROTO_BASE)) {
-      return goPackage.substring(CORE_PROTO_BASE.length()).replace("/", "_");
+    if (!goPackage.startsWith(CORE_PROTO_BASE)) {
+      return proto.getFile().getProto().getPackage().replace(".", "_");
     }
-    return proto.getFile().getProto().getPackage().replace(".", "_");
+    List<String> parts = Arrays.asList(goPackage.split("/"));
+    Collections.reverse(parts);
+    String name = parts.get(0);
+    for (String part : parts) {
```

Comment:
Some brief commentary on what this logic does would be helpful for future eyes. It's not immediately obvious what input corresponds to what output on inspection.